lib/util/dump-collection.ts

```diff
@@ -0,0 +1,337 @@
+
+    if (Object.prototype.toString.call(value) === '[object Object]') {
+      output[newKey] = value;
+      flattenStep(output, value, newKey);
+    }
+
+    output[newKey] = value;
+  }
+}
+
+
+/**
+ * An iteration-order-safe version of lodash.values
+ *
+ * @param object The object containing the values
+ * @param fields The field names to pick in the right order
+ * @returns The values in the same order as the fields
+ * @see https://lodash.com/docs/4.17.15#values
+ */
+function pickValues (object: any, fields: string[]): any[] {
+  return fields.map(f => formatValueForCSV(get(object, f)));
+}
+
+/**
+ * Formats the value for correct CSV output, avoiding to return
+ * values that would badly serialize in CSV.
+ *
+ * @param value The value to format
+ * @returns The value or a string telling the value is not scalar
+ */
+function formatValueForCSV (value: any) {
+  if (isObject(value)) {
+    return '[NOT_SCALAR]';
+  }
+
+  return value;
+}
+
+abstract class AbstractDumper {
+  protected collectionDir: string;
+  protected options: {scroll: string, size: number};
+
+  protected abstract get fileExtension(): string;
+
+  constructor (
+    protected readonly index: string,
+    protected readonly collection: string,
+    protected readonly batchSize: number,
+    protected readonly query: any = {},
+    protected readonly writeStream: stream.Writable,
+  ) {
+    this.options = {
+      scroll: '2s',
+      size: batchSize
+    };
+  }
+
+  /**
+   * One-shot call before the dump. Can be used to
+   * perform setup operations before dumping.
+   *
+   * @returns void
+   */
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  public async setup () {}
+
+  /**
+   * One-shot call before iterating over the data. Can be
+   * used to write the header of the dumped output.
+   */
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  public async writeHeader () {}
+
+  /**
+   * You can put here the logic to write into the dump.
+   *
+   * @param data The data to be written to the dump (can be
+   *             an item or anything else).
+   */
+  abstract writeLine (data: any): Promise<void>
+
+  /**
+   * Iterative call, on each item in the collection to
+   * be dumped. Useful to perform transformations on the data
+   * before writing it in the dump. Usually, writeLine is
+   * called by this hook.
+   *
+   * @param document The document to be written in a line of the dump.
+   */
+  abstract onResult (document: {_id: string, _source: any}): Promise<void>
+
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  public async tearDown () {}
+
+  private async scroll (scrollId: string): Promise<JSONObject> {
+    if (! scrollId) {
+      return null;
+    }
+
+    try {
+      return await global.kuzzle.ask(
+        'core:storage:public:document:scroll',
+        scrollId,
+        { scrollTTL: this.options.scroll }
+      );
+    }
+    catch {
```

Comment:
What if the error is a timeout from the ES cluster for example?

Reply:
As discussed, this should be implemented globally rather than in this PR, see https://github.com/kuzzleio/kuzzle/issues/1659